bvm/ballerina-runtime/src/main/java/io/ballerina/runtime/internal/MapUtils.java

```diff
@@ -82,16 +83,22 @@ public static void handleInherentTypeViolatingRecordUpdate(MapValue mapValue, BS
             // i.e., it is not a `readonly` field or this is an insertion on creation.
             // `initialValue` is only true if this is an update for a field provided in the mapping constructor
             // expression.
-            if (!initialValue && SymbolFlags.isFlagOn(recField.getFlags(), SymbolFlags.READONLY)) {
-
-                throw ErrorCreator.createError(
-                        getModulePrefixedReason(MAP_LANG_LIB, INHERENT_TYPE_VIOLATION_ERROR_IDENTIFIER),
-                        BLangExceptionHelper.getErrorMessage(RuntimeErrors.RECORD_INVALID_READONLY_FIELD_UPDATE,
-                                                             fieldName, recType));
+            if (SymbolFlags.isFlagOn(recField.getFlags(), SymbolFlags.READONLY)) {
+                if (initialValue) {
+                    recFieldType = recField.getFieldType();
+                    if (recFieldType.getTag() == TypeTags.INTERSECTION_TAG) {
+                        recFieldType = ((BIntersectionType) recField.getFieldType()).getConstituentTypes().get(0);
```

Comment:
I don't think this is correct. Does the following still fail?

```ballerina
type Foo record {|
    int[] & readonly x?;
    int[]...; 
|};

public function main() {
    string str = "x";
    int[] arr = [1, 2];
    Foo foo = {[str]: arr};
}
```

Reply:
The above will work as expected since it is not a readonly record. The following will probably fail (won't panic but it should panic).

```ballerina
type Person record {|
    int[] x = [];
|};

public function main() {
    Person & readonly y = {};
}
```

Reply:
I have refactor the fix to address above two cases as well.